File: frontend/src/passguard.ts
import { ContextPasswordChecker } from './context.js'
import { DEFAULT_FRONTEND_BLOCKLIST_TEXT } from './default-blocklist.js'
import { HibpPwnedPasswordClient } from './hibp.js'
import { PasswordBlocklist } from './local-blocklist.js'
import {
  PasswordPolicy,
  type AssessOptions,
  type PasswordPolicyConfig,
} from './password-policy.js'
import type {
  PasswordAssessment,
  PwnedPasswordChecker,
  StrengthEstimator,
} from './types.js'
import { ZxcvbnTsStrengthEstimator } from './zxcvbn-adapter.js'

/** 创建高级 PassGuard 实例时可替换的组件与策略。 */
export interface PassGuardOptions {
  /** 产品名、企业名等所有用户共享的上下文词。 */
  readonly contextWords?: readonly string[]
  /** 自定义本地名单；默认使用包内置的 25,000 条高频名单。 */
  readonly blocklist?: PasswordBlocklist
  /** 设为 false 可关闭强度估算；默认使用 zxcvbn-ts。 */
  readonly strengthEstimator?: StrengthEstimator | false
  /** 设为 false 可关闭 HIBP；默认仅发送 SHA-1 前 5 位做 k-anonymity 查询。 */
  readonly pwnedChecker?: PwnedPasswordChecker | false
  /** 默认策略的局部覆盖；未提供字段继续使用安全默认值。 */
  readonly config?: Partial<PasswordPolicyConfig>
}

/** 单次 PassGuard.check 调用选项。 */
export interface PassGuardCheckOptions extends Partial<AssessOptions> {
  /** 是否使用 MFA 场景最小长度；默认 false。 */
  readonly mfaProtected?: boolean
}

/**
 * PassGuard 的高级 API。实例是无状态的，可在整个应用中安全复用。
 */
export class PassGuard {
  readonly #policy: PasswordPolicy

  /**
   * 创建高级实例并加载缺省组件。
   *
   * @param options 名单、上下文词、估算器、泄露检查器和策略覆盖
   */
  constructor(options: PassGuardOptions = {}) {
    const blocklist = options.blocklist
      ?? PasswordBlocklist.fromText(DEFAULT_FRONTEND_BLOCKLIST_TEXT)
    const strengthEstimator = options.strengthEstimator === false
      ? undefined
      : (options.strengthEstimator ?? new ZxcvbnTsStrengthEstimator())
    const pwnedChecker = options.pwnedChecker === false
      ? undefined
      : (options.pwnedChecker ?? new HibpPwnedPasswordClient())

    this.#policy = new PasswordPolicy({
      blocklist,
      contextChecker: new ContextPasswordChecker(options.contextWords),
      ...(strengthEstimator ? { strengthEstimator } : {}),
      ...(pwnedChecker ? { pwnedChecker } : {}),
      ...(options.config ? { config: options.config } : {}),
    })
  }

  /**
   * 完整检查密码。
   *
   * @param password 原始密码；执行 NFC 规范化但不会 trim
   * @param options MFA、上下文和取消选项
   * @returns 只读评估结果
   */
  check(password: string, options: PassGuardCheckOptions = {}): Promise<PasswordAssessment> {
    return this.#policy.assess(password, {
      mfaProtected: options.mfaProtected ?? false,
      ...(options.context ? { context: options.context } : {}),
      ...(options.signal ? { signal: options.signal } : {}),
    })
  }
}

/** 一行创建可复用的完整密码防御实例。 */
export function createPassGuard(options: PassGuardOptions = {}): PassGuard {
  return new PassGuard(options)
}

File: java/src/main/java/dev/flyfish/passguard/PassGuard.java
package dev.flyfish.passguard;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

/**
 * PassGuard 的高级入口。实例是无状态的，可在整个应用中安全复用。
 */
public final class PassGuard {
    /** 内置后端弱密码名单的 classpath 资源路径。 */
    public static final String DEFAULT_BLOCKLIST_RESOURCE =
            "/weak-passwords/backend-blocklist.txt";

    private final PasswordPolicy policy;

    private PassGuard(PasswordPolicy policy) {
        this.policy = policy;
    }

    /**
     * 使用内置 125,691 条名单、nbvcxz 和 HIBP 创建完整防御实例。
     */
    public static PassGuard create() {
        return builder().build();
    }

    /**
     * 创建不访问外部网络的实例，适合离线任务或低延迟本地预检。
     */
    public static PassGuard localOnly() {
        return builder().disablePwnedCheck().build();
    }

    /** @return 新的高级配置构建器 */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * 使用单因素最小长度和空上下文检查密码。
     *
     * @param password 原始密码
     * @return 评估结果
     */
    public PasswordAssessment check(String password) {
        return check(password, false, PasswordContext.empty());
    }

    /**
     * 使用空上下文检查密码。
     *
     * @param password 原始密码
     * @param mfaProtected 是否使用 MFA 场景最小长度
     * @return 评估结果
     */
    public PasswordAssessment check(String password, boolean mfaProtected) {
        return check(password, mfaProtected, PasswordContext.empty());
    }

    /**
     * 使用完整用户与业务上下文检查密码。
     *
     * @param password 原始密码
     * @param mfaProtected 是否使用 MFA 场景最小长度
     * @param context 用户与业务上下文；可为 {@code null}
     * @return 评估结果
     */
    public PasswordAssessment check(
            String password,
            boolean mfaProtected,
            PasswordContext context
    ) {
        return policy.assess(password, mfaProtected, context);
    }

    /**
     * {@link PassGuard} 的可复用组件构建器。
     */
    public static final class Builder {
        private PasswordPolicyConfig config = PasswordPolicyConfig.secureDefaults();
        private LocalBlocklist blocklist;
        private final List<String> contextWords = new ArrayList<>();
        private StrengthEstimator strengthEstimator;
        private boolean defaultStrengthEstimator = true;
        private PwnedPasswordChecker pwnedChecker;
        private boolean defaultPwnedChecker = true;

        private Builder() {}

        /**
         * @param config 完整策略配置
         * @return 当前构建器
         */
        public Builder config(PasswordPolicyConfig config) {
            this.config = Objects.requireNonNull(config, "config");
            return this;
        }

        /**
         * @param blocklist 自定义本地名单
         * @return 当前构建器
         */
        public Builder blocklist(LocalBlocklist blocklist) {
            this.blocklist = Objects.requireNonNull(blocklist, "blocklist");
            return this;
        }

        /**
         * 累积产品名、企业名等全局上下文词。
         *
         * @param words 全局词
         * @return 当前构建器
         */
        public Builder contextWords(String... words) {
            Objects.requireNonNull(words, "words");
            return contextWords(Arrays.asList(words));
        }

        /**
         * 累积产品名、企业名等全局上下文词。
         *
         * @param words 全局词集合
         * @return 当前构建器
         */
        public Builder contextWords(Collection<String> words) {
            Objects.requireNonNull(words, "words");
            for (String word : words) {
                if (!ContextPasswordChecker.isBlank(word)) contextWords.add(word);
            }
            return this;
        }

        /**
         * @param estimator 自定义强度估算器
         * @return 当前构建器
         */
        public Builder strengthEstimator(StrengthEstimator estimator) {
            this.strengthEstimator = Objects.requireNonNull(estimator, "estimator");
            this.defaultStrengthEstimator = false;
            return this;
        }

        /**
         * 关闭强度估算。
         *
         * @return 当前构建器
         */
        public Builder disableStrengthEstimator() {
            this.strengthEstimator = null;
            this.defaultStrengthEstimator = false;
            return this;
        }

        /**
         * @param checker 自定义泄露密码检查器
         * @return 当前构建器
         */
        public Builder pwnedChecker(PwnedPasswordChecker checker) {
            this.pwnedChecker = Objects.requireNonNull(checker, "checker");
            this.defaultPwnedChecker = false;
            return this;
        }

        /**
         * 关闭远程泄露检查。
         *
         * @return 当前构建器
         */
        public Builder disablePwnedCheck() {
            this.pwnedChecker = null;
            this.defaultPwnedChecker = false;
            return this;
        }

        /**
         * 加载缺省组件并创建无状态实例。
         *
         * @return 可复用的 PassGuard
         * @throws IllegalStateException 内置名单加载失败
         */
        public PassGuard build() {
            LocalBlocklist effectiveBlocklist = blocklist == null
                    ? loadDefaultBlocklist()
                    : blocklist;
            StrengthEstimator effectiveStrengthEstimator = defaultStrengthEstimator
                    ? new NbvcxzStrengthEstimator()
                    : strengthEstimator;
            PwnedPasswordChecker effectivePwnedChecker = defaultPwnedChecker
                    ? new HibpPwnedPasswordClient()
                    : pwnedChecker;
            PasswordPolicy effectivePolicy = new PasswordPolicy(
                    config,
                    effectiveBlocklist,
                    new ContextPasswordChecker(contextWords),
                    effectiveStrengthEstimator,
                    effectivePwnedChecker
            );
            return new PassGuard(effectivePolicy);
        }

        private static LocalBlocklist loadDefaultBlocklist() {
            try {
                return LocalBlocklist.fromClasspath(DEFAULT_BLOCKLIST_RESOURCE);
            } catch (IOException exception) {
                throw new IllegalStateException(
                        "PassGuard 内置弱密码名单加载失败", exception);
            }
        }
    }
}
